src/librustc/traits/coherence.rs

```diff
@@ -504,6 +504,15 @@ fn ty_is_local_constructor(ty: Ty<'_>, in_crate: InCrate) -> bool {
 
         ty::Adt(def, _) => def_id_is_local(def.did, in_crate),
         ty::Foreign(did) => def_id_is_local(did, in_crate),
+        ty::Opaque(did, _) => {
+            // Check the underlying type that this opaque
+            // type resolves to.
+            // This recursion will eventually terminate,
+            // since we've already managed to successfully
+            // resolve all opaque types by this point
+            let real_ty = tcx.type_of(did);
+            ty_is_local_constructor(tcx, real_ty, in_crate)
```

Comment:
I'm not sure if this is the only interpretation that makes sense. In fact, it's a bit surprising to me, since it means that impls can "peer into" the hidden type of an `impl Trait`, albeit only in limited ways. On the one hand, if the impl is within the scope of the `impl Trait`, this seems logical enough. On the other hand, if the impl is *outside* the scope of the `impl Trait`, it seems less clear.

As you say, the restrictions against having an opaque type in an impl somewhat prevent this from coming up -- but I guess the bottom line is that they are not complete, which is why it *is* coming up. I can't really come up with a strong reason why `impl Opaque` is disallowed by `impl Foo<Opaque>` is *allowed* -- whatever hard questions we have to answer about overlap seem to arise just as easily in both cases.

Another option would be to prohibit `Opaque` types more deeply (though we have to think about the interaction of that with type aliases and associated types).  Even then though, similar questions could arise because of auto traits -- well, it's not quite the same I guess, since we are asking 

I'd be somewhat tempted to say that `Opaque` is always considered *non-local* -- have you tried that? It would mean that we can't assume negative things about `Opaque` .. 

Well, I realize now there is one additional complication, which is fundamental traits. If we don't do coherence in a "revealed" context, we could make some mistakes. For example, we might determine that `impl Clone` does *not* implement `FnOnce()`, but we don't actually *know* that. This would happen because we would attempt to prove `impl Clone: FnOnce()` and fail and interpret that failure as "does not implement".

To solve that, I guess, we could *either* do coherence in a revealed context (where we strongly normalize opaque types, and hence they don't exist) or we could make a smarter form of negation (more like what chalk has, which sidesteps this trap). Basically it would know that `not { impl Clone: FnOnce() }` is not provable. I'd have to check back in the code to see how easy this would be to add.

Reply:
That's a good point - I hadn't considered the `impl Foo<Opaque>` case.

I think it makes sense to say that opaque types are always non-local with respect to coherence. Otherwise, you could run into a situation where changing the underlying type is a breaking change (e.g. `impl Foo<Opaque>` no longer references any local types), which is exactly what opaque types are supposed to help prevent.